### ISSUE TITLE
4 Stages

### ISSUE THREAD

**Issue by psaegert** [psaegert/flash-ansr#20]
# Motivation
- Successful applications of Reinforcement Learning (RL) training stages in both NLP ([OpenAI o1](https://openai.com/index/openai-o1-system-card/) among others) and SR ([DSO](https://github.com/dso-org/deep-symbolic-optimization))

# Proposal
Add additional training steps after pre-training, in particular fine-Tuning on own predictions with RL using the log FVU as a reward ([REINFORCE](https://paperswithcode.com/method/reinforce) or Risk-Seeking as in [DSR](https://arxiv.org/abs/1912.04871))
- [ ] On entire training set and permanent
- [ ] Specialized on instance at hand and temporary (perhaps faster than DSO because model is already pre-trained)

# Implementation
## 1. Pre-Training
Use existing pre-training

## 2. RL-Fine-Tuning
- [ ] Implement [REINFORCE](https://paperswithcode.com/method/reinforce)
- [ ] Implement Risk-Seeking Policy Gradient as in [DSR](https://arxiv.org/abs/1912.04871
- [ ] Implement RL-training module for large scale general fine-tuning on entire training distribution

## 3. RL-"Specialization"
- [ ] Use RL-training module for instance-level fine-tuning on given data

Usage could look like this:
```python
FlashANSR.specialize(X, y, generation_type='softmax', iterations=...)
```

## 4. Inference
Use existing inference:
```python
FlashANSR.fit(X, y, generation_type='beam_search', ...)
```
